### PR TITLE
AArch64: Enable NUMA for FDT

### PR DESCRIPTION
**Pull request by MrXinWang**
This PR will make `--numa` and related config from Cloud Hypervisor command line work on AArch64 platform with FDT.

### CODE REVIEW

File: docs/memory.md [hunk 1]
```diff
@@ -439,7 +439,9 @@ workload run more efficiently.
 
 Multiple values can be provided to define the list. Each value is a string
 referring to an existing memory zone identifier. Values are separated from
-each other with the `:` separator.
+each other with the `:` separator. Note that the memory zone id used by
+different NUMA nodes should also be different, i.e. this is not allowed:
+`--numa guest_numa_id=0,memory_zones=mem0 guest_numa_id=1,memory_zones=mem0`
```
Comment (MrXinWang) [inline docs/memory.md:444]: Hi @sboeuf, could you please confirm this change is correct? As from my understanding, different NUMA id cannot share the same memory zone...If this is correct, should we also add some code to forbid user setting command line like this? Thanks!

File: vmm/src/vm.rs [hunk 2]
```diff
@@ -651,7 +646,6 @@ impl Vm {
         })
     }
 
-    #[cfg(feature = "acpi")]
```
Comment (michael2012z) [inline vmm/src/vm.rs:654]: @rbradford , is there any problem on X86 if we make NUMA not specific to ACPI?

Comment (rbradford) [reply vmm/src/vm.rs:654]: Should be fine, the `--numa` parsing is not feature gated.

**Review by rbradford — APPROVED** [review]
You have a rubber stamp from me but I think @sboeuf should answer the question about zones since he is the expert there.

**Comment by MrXinWang** [comment]
> You have a rubber stamp from me but I think @sboeuf should answer the question about zones since he is the expert there.

of course :))

File: vmm/src/vm.rs [hunk 3]
```diff
@@ -519,7 +518,7 @@ pub struct Vm {
     vm: Arc<dyn hypervisor::Vm>,
     #[cfg(all(feature = "kvm", target_arch = "x86_64"))]
     saved_clock: Option<hypervisor::ClockData>,
-    #[cfg(feature = "acpi")]
+    #[allow(dead_code)]
```
Comment (sboeuf) [inline vmm/src/vm.rs:479]: Is it the way we want to handle this? The alternative would be to provide the features for which this is relevant.

File: docs/memory.md (hunk 1 above)
Comment (sboeuf) [reply docs/memory.md:444]: Yes you're totally right. We can't have the same memory zone belonging to more than one NUMA node. If this is not validated in the code, then yes we should add a check to prevent the user from specifying such configuration.

Comment (sboeuf) [reply docs/memory.md:444]: I'd rephrase with:
```
Note that a memory zone must belong to a single NUMA node. The following
configuration is incorrect, therefore not allowed:
`--numa guest_numa_id=0,memory_zones=mem0 guest_numa_id=1,memory_zones=mem0`
```

File: tests/integration.rs [hunk 4]
```diff
@@ -2337,6 +2337,63 @@ mod tests {
             handle_child_output(r, &output);
         }
 
+        #[test]
+        #[cfg(target_arch = "aarch64")]
+        fn test_aarch64_guest_numa_nodes_dt() {
```
Comment (sboeuf) [inline tests/integration.rs:2342]: Why do you need a dedicated test?
You should be able to reuse `test_guest_numa_nodes()`.
I think you should reuse `test_guest_numa_nodes()` to make it a common function that is called from `test_guest_numa_nodes_acpi()` and `test_guest_numa_nodes_dt()`. Of course you would apply the correct feature gating on top of each to make sure they can run as expected.

File: vmm/src/vm.rs (hunk 3 above)
Comment (MrXinWang) [reply vmm/src/vm.rs:479]: Yeah (more or less), as the clippy will also check `--features=kvm` without `acpi` on x86_64, and in this case this `numa_nodes: NumaNodes` field is never used. The `#[allow(dead_code)]` is intended just to mute clippy.

On arm everything is fine (both `--features=kvm` and `--features=kvm, acpi` will use this field).

It is kinda tricky to provide a proper gate here...

File: tests/integration.rs (hunk 4 above)
Comment (MrXinWang) [reply tests/integration.rs:2342]: Ah, this is because we want to make sure that both the DT and ACPI will be tested on Arm. For now, we only build the Cloud Hypervisor for once in CI using `cargo build`. Using feature gates cannot cover both cases.

But I like the idea of reusing `test_guest_numa_nodes()` as a common function. Let me do some refactoring :)) Thank you very much for the suggestion!

File: docs/memory.md (hunk 1 above)
Comment (MrXinWang) [reply docs/memory.md:444]: > Yes you're totally right. We can't have the same memory zone belonging to more than one NUMA node. If this is not validated in the code, then yes we should add a check to prevent the user from specifying such configuration.

AFAIK the current code is not validating that - As I remember I tried this once...

Then I will add some checking code. How about this? @sboeuf

Here (https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/vmm/src/config.rs#L1689) when we parsing the config, we add a global/static data structure, for example a vector, to store the already used memory zones. If we meet a used memory zone when parsing the rest of NUMA nodes, we raise an error.

**Comment by MrXinWang** [comment]
Temporary making this PR as draft to fix some useful comments.

File: tests/integration.rs (hunk 4 above)
Comment (rbradford) [reply tests/integration.rs:2342]: Can you can test the two code paths by firmware boot vs direct boot (acpi vs fdt)?

Comment (MrXinWang) [reply tests/integration.rs:2342]: > Can you can test the two code paths by firmware boot vs direct boot (acpi vs fdt)?

Hi @rbradford ! Not very sure what you mean (as there is a typo), but yeah, the idea is just to change the path of `--kernel`. However on arm we need to run both of two cases, i.e. `--kernel=direct_kernel_boot_path()` once and `--kernel=edk2_path()` once. I don't think this is possible in one single test case unless using a loop. So I would extract most of the common code into a helper function and split the original test case into `test_guest_numa_nodes_acpi()` and `test_guest_numa_nodes_dt()`

Comment (rbradford) [reply tests/integration.rs:2342]: This is exactly what I was thinking.

**Comment by MrXinWang** [comment]
Hi @sboeuf @rbradford ! I believe I have addressed all of your concerns:

- https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2950/commits/2e925d6c67f8fb3bbeafbca927c195401b47f42a and https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2950/commits/27401b1765688b41b751e70b4ed522f251698cee add proper gates to the NUMA related data structures and functions.
- https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2950/commits/2dc1ed271d11b3d9629123e1ddd6db7ee75a6f98 adds validation code to `--numa` memory nodes, an example is given below if `mem1` is used twice:
```
Error parsing config: Error validating configuration: Memory zone: mem1 belongs to multiple NUMA nodes
```
- https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2950/commits/240ac48ad58690b036630463afd985e6a58e5ef0 extracts most of the common code into a helper function and split the original test case into `test_guest_numa_nodes_acpi()` and `test_guest_numa_nodes_dt()`

File: arch/src/lib.rs [hunk 5]
```diff
@@ -96,6 +101,46 @@ fn pagesize() -> usize {
...
+    pub memory_regions: Vec<Arc<GuestRegionMmap>>,
+    pub hotplug_regions: Vec<Arc<GuestRegionMmap>>,
+    pub cpus: Vec<u8>,
+    pub distances: BTreeMap<u32, u8>,
+    pub memory_zones: Vec<String>,
+    #[cfg(target_arch = "x86_64")]
+    pub sgx_epc_sections: Vec<SgxEpcSection>,
+}
+
+impl NumaNode {
```
Comment (rbradford) [inline arch/src/lib.rs:115]: What's the benefit of having all these accessors if all the fields are pub?

File: vmm/src/config.rs [hunk 6]
```diff
@@ -1915,6 +1920,20 @@ impl VmConfig {
...
         }
+
+        if let Some(numa) = &self.numa {
+            let mut used_numa_node_memory_zones = Vec::new();
+            for numa_node in numa.iter() {
+                for memory_zone in numa_node.memory_zones.clone().unwrap().iter() {
+                    if !used_numa_node_memory_zones.contains(memory_zone) {
+                        used_numa_node_memory_zones.push(memory_zone.to_string());
+                    } else {
+                        return Err(ValidationError::MemoryZoneReused(memory_zone.to_string()));
```
Comment (rbradford) [inline vmm/src/config.rs:1931]: This is nice, but could your also give the conficting node ids to make it easier for the user to find their conflict?

Comment (MrXinWang) [reply vmm/src/config.rs]: Done. Yes that makes even more sense :))

File: arch/src/lib.rs (hunk 5 above)
Comment (MrXinWang) [reply arch/src/lib.rs:115]: To be honest I don't know....I didn't think too much when moving this part and just moved the original code as it is.

Maybe @sboeuf is more experienced to answer this question?

Comment (sboeuf) [reply arch/src/lib.rs:115]: Check if the code is doing extensive use of directly accessing the fields. If not, then try to make sure it relies on accessors only.

**Review by sboeuf — COMMENTED** [review]
Integration test looks good, thanks for fixing it!

File: arch/src/lib.rs (hunk 5 above)
Comment (rbradford) [reply arch/src/lib.rs:115]: I think you can remove the accessors if all the fields are pub.

Comment (MrXinWang) [reply arch/src/lib.rs]: Hi @rbradford. Yes and done. Such change is reflected in https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2950/commits/fdd0a6c469b638ee5c79ee059a71bb9d2f1c578c

**Review by sboeuf — APPROVED** [review]
Great! Thanks for addressing all the comments and issues!
